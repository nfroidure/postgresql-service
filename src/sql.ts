export const SQLStatementTypeSymbol = Symbol('SQLStatement');
export const SQLPartTypeSymbol = Symbol('SQLPart');
export const SQLValuesTypeSymbol = Symbol('SQLValues');
export const SQLTypesSymbols = [
  SQLPartTypeSymbol,
  SQLStatementTypeSymbol,
] as const;

export type SQLStatement = {
  type: typeof SQLStatementTypeSymbol;
  parts: string[];
  values: SQLValue[];
  text: string;
};

export type SQLPart = {
  type: typeof SQLPartTypeSymbol;
  parts: string[];
  values: SQLValue[];
};

export type SQLValues = {
  type: typeof SQLValuesTypeSymbol;
  values: SQLValue[];
};

export type SQLValue = string | number | boolean | object | null | undefined;

export type SQLStringLiteralParameter =
  | SQLStatement
  | SQLPart
  | SQLValue
  | SQLValues;

export const joinSQLValues = (values: SQLValue[]): SQLValues => {
  return {
    type: SQLValuesTypeSymbol,
    values,
  };
};

export const createSQLPart = (str: string): SQLPart => {
  return {
    type: SQLPartTypeSymbol,
    parts: [str],
    values: [],
  };
};

export const emptySQLPart = createSQLPart('');

export const escapeSQLIdentifier = (str: string): SQLPart => {
  return createSQLPart('"' + str.replace(/"/g, '""') + '"');
};

function isSQLValues(value: SQLStringLiteralParameter): value is SQLValues {
  return (
    typeof value === 'object' &&
    value !== null &&
    ((value as unknown) as SQLValues).type == SQLValuesTypeSymbol
  );
}

function isSQLPart(value: SQLStringLiteralParameter): value is SQLPart {
  return (
    typeof value === 'object' &&
    value !== null &&
    ((value as unknown) as SQLPart).type == SQLPartTypeSymbol
  );
}

function isSQLStatement(
  value: SQLStringLiteralParameter,
): value is SQLStatement {
  return (
    typeof value === 'object' &&
    value !== null &&
    ((value as unknown) as SQLStatement).type == SQLStatementTypeSymbol
  );
}

/* Architecture Note #1.1: Tagged template queries

The `pg` module uses simple `$n` placeholder for queries values
 that are provided in an array.

This tagged template function adds a level of abstraction
 transforming the following expression:
 `sql\`SELECT * FROM users WHERE id=${userId}\`` with `userId=1` }
 into `{ text: 'SELECT * FROM users WHERE id=$1', values: [1] }`
 under the hood.
*/
export default function sql<T extends SQLStringLiteralParameter[]>(
  chunks: TemplateStringsArray,
  ...parameters: T
): SQLStatement {
  const { parts, values } = mergeSQLChunks(chunks, parameters);
  return {
    type: SQLStatementTypeSymbol,
    parts,
    values,
    get text() {
      return this.parts.reduce(
        (text: string, part: string, index: number) =>
          text + part + (index < this.values.length ? '$' + (index + 1) : ''),
        '',
      );
    },
  };
}

export function sqlPart<T extends SQLStringLiteralParameter[]>(
  chunks: TemplateStringsArray,
  ...parameters: T
): SQLPart {
  const { parts, values } = mergeSQLChunks(chunks, parameters);
  return {
    type: SQLPartTypeSymbol,
    parts,
    values,
  };
}

function mergeSQLChunks(chunks, parameters) {
  const parts = [];
  const values: SQLValue[] = [];
  const parametersLength = parameters.length;

  for (let i = 0, len = chunks.length; i < len; i++) {
    const unterminatedPart = parts.length > values.length;

    if (i >= parametersLength) {
      if (unterminatedPart) {
        parts.push((parts.pop() || '') + chunks[i]);
      } else {
        parts.push(chunks[i]);
      }
      continue;
    }

    if (isSQLStatement(parameters[i]) || isSQLPart(parameters[i])) {
      if (unterminatedPart) {
        parts.push(
          (parts.pop() || '') +
            chunks[i] +
            (parameters[i] as SQLStatement).parts[0],
        );
      } else {
        parts.push(chunks[i] + (parameters[i] as SQLStatement).parts[0]);
      }
      parts.push(...(parameters[i] as SQLStatement).parts.slice(1));
      values.push(...(parameters[i] as SQLStatement).values);
    } else if (isSQLValues(parameters[i])) {
      if (!unterminatedPart) {
        parts.push((parts.pop() || '') + chunks[i]);
      } else {
        parts.push(chunks[i]);
      }
      parts.push(
        ...(parameters[i] as SQLValues).values.slice(1).map(() => ', '),
      );
      values.push(...(parameters[i] as SQLValues).values);
    } else {
      if (unterminatedPart) {
        parts.push((parts.pop() || '') + chunks[i]);
      } else {
        parts.push(chunks[i]);
      }
      values.push(parameters[i]);
    }
  }

  return {
    parts,
    values,
  };
}
