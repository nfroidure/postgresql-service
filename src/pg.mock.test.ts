import assert from 'assert';
import YError from 'yerror';
import initPGMock from './pg.mock';
import sql from './sql';

describe('initPGMock', () => {
  test('should work', async () => {
    const pg = await initPGMock();

    assert(pg);
  });

  describe('pg.query', () => {
    test('should work', async () => {
      const { service: pg } = await initPGMock();

      pg.query.mockResultOnce({ rows: [], fields: {} });

      await pg.query(sql`SELECT * FROM users WHERE id = ${1}`);

      expect(pg.query.mock.calls).toMatchSnapshot();
    });

    test('should fail with malformed query', async () => {
      const { service: pg } = await initPGMock();

      pg.query.mockResultOnce({ rows: [], fields: {} });

      try {
        await pg.query(sql`SELECT * FROM`);
        throw new YError('E_UNEXPECTED_SUCCESS');
      } catch (err) {
        expect({
          errorCode: err.code,
          errorParams: err.params,
        }).toMatchSnapshot();
      }
    });
  });

  describe('pg.queries', () => {
    test('should work', async () => {
      const { service: pg } = await initPGMock();

      pg.queries.mockResultOnce([{ rows: [], fields: {} }]);

      await pg.queries([
        sql`SELECT * FROM users WHERE id = ${1}`,
        sql`SELECT * FROM users WHERE id = ${1}`,
      ]);

      expect(pg.queries.mock.calls).toMatchSnapshot();
    });
  });

  describe('pg.transaction', () => {
    test('should work', async () => {
      const { service: pg } = await initPGMock();

      pg.transaction.mockResultOnce([{ rows: [], fields: {} }]);

      await pg.transaction([
        sql`SELECT * FROM users WHERE id = ${1}`,
        sql`SELECT * FROM users WHERE id = ${1}`,
      ]);

      expect(pg.transaction.mock.calls).toMatchSnapshot();
    });
  });
});
