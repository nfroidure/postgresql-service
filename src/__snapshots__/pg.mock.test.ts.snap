// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initPGMock pg.queries should work 1`] = `
Array [
  Array [
    Array [
      Object {
        "parts": Array [
          "SELECT * FROM users WHERE id = ",
          "",
        ],
        "text": "SELECT * FROM users WHERE id = $1",
        "type": Symbol(SQLStatement),
        "values": Array [
          1,
        ],
      },
      Object {
        "parts": Array [
          "SELECT * FROM users WHERE id = ",
          "",
        ],
        "text": "SELECT * FROM users WHERE id = $1",
        "type": Symbol(SQLStatement),
        "values": Array [
          1,
        ],
      },
    ],
  ],
]
`;

exports[`initPGMock pg.query should fail with malformed query 1`] = `
Object {
  "errorCode": "E_INVALID_QUERY",
  "errorParams": Array [
    "SELECT * FROM",
    [Error: syntax error at end of input],
  ],
}
`;

exports[`initPGMock pg.query should work 1`] = `
Array [
  Array [
    Object {
      "parts": Array [
        "SELECT * FROM users WHERE id = ",
        "",
      ],
      "text": "SELECT * FROM users WHERE id = $1",
      "type": Symbol(SQLStatement),
      "values": Array [
        1,
      ],
    },
  ],
]
`;

exports[`initPGMock pg.transaction should work 1`] = `
Array [
  Array [
    Array [
      Object {
        "parts": Array [
          "SELECT * FROM users WHERE id = ",
          "",
        ],
        "text": "SELECT * FROM users WHERE id = $1",
        "type": Symbol(SQLStatement),
        "values": Array [
          1,
        ],
      },
      Object {
        "parts": Array [
          "SELECT * FROM users WHERE id = ",
          "",
        ],
        "text": "SELECT * FROM users WHERE id = $1",
        "type": Symbol(SQLStatement),
        "values": Array [
          1,
        ],
      },
    ],
  ],
]
`;
